src/gpio.rs

```diff
@@ -3,13 +3,113 @@
 //! concrete pin type.
 
 use core::convert::Infallible;
+use core::future::Future;
 use core::marker::PhantomData;
+use core::pin::Pin as FuturePin;
+use core::sync::atomic::{AtomicU32, Ordering};
+use core::task::{Context, Poll};
 
-use embassy_hal_internal::{Peri, PeripheralType};
+use embassy_sync::waitqueue::AtomicWaker;
+use embassy_hal_internal::{Peri, PeripheralType, interrupt::InterruptExt};
 use paste::paste;
 
+use crate::pac::interrupt;
 use crate::pac::port0::pcr0::{Dse, Inv, Mux, Pe, Ps, Sre};
 
+struct BitIter(u32);
+
+impl Iterator for BitIter {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.0.trailing_zeros() {
+            32 => None,
+            b => {
+                self.0 &= !(1 << b);
+                Some(b as usize)
+            }
+        }
+    }
+}
+
+const PORT_COUNT: usize = 5;
+
+static WAKERS: [AtomicWaker; PORT_COUNT] = [const { AtomicWaker::new() }; PORT_COUNT];
+static INTERRUPT_DETECTED: [AtomicU32; PORT_COUNT] = [const { AtomicU32::new(0) }; PORT_COUNT];
+/// Interrupt trigger levels.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum InterruptTrigger {
+    LevelLow,
+    RisingEdge,
+    FallingEdge,
+    AnyEdges,
+    LevelHigh,
+}
+
+pub(crate) unsafe fn init() {
+    interrupt::GPIO0.enable();
+    interrupt::GPIO1.enable();
+    interrupt::GPIO2.enable();
+    interrupt::GPIO3.enable();
+    interrupt::GPIO4.enable();
+
+    cortex_m::interrupt::enable();
+}
+
+#[cfg(feature = "rt")]
+fn irq_handler(port_index: usize, gpio_base: *const crate::pac::gpio0::RegisterBlock) {
+    let gpio = unsafe { &*gpio_base };
+    let isfr = gpio.isfr0().read().bits();
+
+    if isfr != 0 {
+        // Clear all pending interrupts
+        gpio.isfr0().write(|w| unsafe { w.bits(isfr) });
+
+        // Wake the corresponding port waker
+        if port_index < WAKERS.len() {
+            WAKERS[port_index].wake();
+        }
+
+        // Disable all pin interrupts that fired to prevent re-triggering
+        for pin in BitIter(isfr) {
+            gpio.icr(pin).modify(|_, w| w.irqc().irqc0()); // Disable interrupt
+        }
+
+        INTERRUPT_DETECTED[port_index].fetch_or(isfr, Ordering::Relaxed);
+    }
```

Comment:
because of the `BitIter` usage, we can modify this slightly:

```suggestion
    for pin in BitIter(isfr) {
        // Clear all pending interrupts
        gpio.isfr0().write(|w| unsafe { w.bits(1 << pin) });
        gpio.icr(pin).modify(|_, w| w.irqc().irqc0()); // Disable interrupt

        INTERRUPT_DETECTED[port_index].fetch_or(isfr, Ordering::Relaxed);

        // Wake the corresponding port waker
        if port_index < WAKERS.len() {
            WAKERS[port_index].wake();
        }
    }
```

Reply:
Done, added commit. Thanks for the suggestion